cmd/build.go

```diff
@@ -144,21 +144,18 @@ func runBuild(cmd *cobra.Command, _ []string, newClient ClientFactory) (err erro
 		function.Image = config.Image
 	}
 
-	// All set, let's write changes in the config to the disk
-	err = function.Write()
-	if err != nil {
-		return
-	}
-
-	// if registry was not changed via command line flag meaning it's empty
-	// keep the same registry by setting the config.registry to empty otherwise
-	// trust viper to override the env variable with the given flag if both are specified
-	if regFlag, _ := cmd.Flags().GetString("registry"); regFlag == "" {
-		config.Registry = ""
-	}
-
 	// Choose a builder based on the value of the --builder flag
 	var builder fn.Builder
+	if config.Builder == "" {
+		if function.Builder == "" {
+			config.Builder, function.Builder = "pack", "pack"
+		} else {
+			config.Builder = function.Builder
+		}
+	} else {
+		function.Builder = config.Builder
+	}
+
```

Comment:
The command instance allows us to check if a flag was actually provided, so perhaps this will suffice:

```go
  if f.Builder == "" || cmd.Flags().Changed("builder") {
    f.Builder = config.Builder
    if err = f.Write(); err != nil {
      return
    }
  }
```

Reply:
Thanks, this looks much better!